### PR TITLE
Test calculation output shape for Broadcast op, relax restrictions for partially dynamic input data 

### PR DESCRIPTION
**Pull request by mbencer**
Tested combinations from CVS-31915

### CODE REVIEW

File: ngraph/test/type_prop/broadcast.cpp [hunk 1]
```diff
@@ -439,6 +439,301 @@ TYPED_TEST_P(BroadcastTests, broadcast_axes_et_wrong)
...
+}
+
+TYPED_TEST_P(BroadcastTests, broadcast_explicit_target_shape_static_rank)
+{
+    const auto data = make_shared<op::Parameter>(element::f32, PartialShape::dynamic());
+    const auto target_shape = make_shared<op::Parameter>(element::i64, PartialShape::dynamic(1));
+    const auto axes_mapping = make_shared<op::Parameter>(element::i64, PartialShape::dynamic());
+
+    auto bc = make_shared<TypeParam>(data, target_shape, axes_mapping, "EXPLICIT");
+    ASSERT_TRUE(bc->get_output_partial_shape(0).rank().is_dynamic());
```
[the remark this answers is not in the document]
Comment (mbencer) [reply ngraph/test/type_prop/broadcast.cpp:467]: In this case only rank of `target_shape` is known. To have `output_rank` known `target_shape` should be `static` - for example `target_shape={5}` means that `output_rank=5`

File: ngraph/src/ngraph/op/util/broadcast_base.cpp [hunk 2]
```diff
@@ -138,13 +150,27 @@ void op::util::BroadcastBase::validate_and_infer_types()
     }
 
     PartialShape result_shape{PartialShape::dynamic()};
-    auto input_rank = input_value(0).get_partial_shape().rank();
-    auto output_rank = input_value(1).get_partial_shape();
-    if (input_rank.is_static() && output_rank.is_static() && output_rank[0].is_static())
+    const auto input_rank = input_value(0).get_partial_shape().rank();
+    const auto output_shape = input_value(1).get_partial_shape();
+    const bool is_output_rank_static =
+        output_shape.rank().is_static() && output_shape[0].is_static();
```
[the remark this answers is not in the document]
Comment (mbencer) [reply ngraph/src/ngraph/op/util/broadcast_base.cpp]: right, I changed it to `is_target_shape_known`

File: ngraph/src/ngraph/op/util/broadcast_base.cpp [hunk 3]
```diff
@@ -204,9 +230,9 @@ void op::util::BroadcastBase::validate_and_infer_types()
     }
     else if (m_mode.m_type == BroadcastType::NUMPY || m_mode.m_type == BroadcastType::PDPD)
     {
-        if (get_input_partial_shape(0).is_static() && get_input_partial_shape(1).is_static())
+        if (get_input_partial_shape(0).rank().is_static() && get_input_partial_shape(1).is_static())
```
[the remark this answers is not in the document]
Comment (mbencer) [reply ngraph/src/ngraph/op/util/broadcast_base.cpp]: done

**Comment by mbencer** [comment]
> I haven't looked through all tests. However after review of spreadsheet with matrix of cases we came to conclusion we should clarify the resulting shapes.
> My understanding is that, we should leverage as much as possible information available from `target_shape` input in case of `numpy` and `explicit` broadcasting rules, which I'd like to emphasize are **uni-directional** modes.

You are right. I've changed rules for `numpy`. Now if `target_shape` is constant, `output_shape` is always set (validation is realized if possible - if input data dimensions are known)

File: ngraph/test/type_prop/broadcast.cpp [hunk 4]
```diff
@@ -439,6 +439,359 @@ TYPED_TEST_P(BroadcastTests, broadcast_axes_et_wrong)
...
+
+    // const axes mapping
+    const auto axes_mapping_const =
+        op::Constant::create(element::i64, Shape{3}, vector<int64_t>{0, 2, 1});
+    bc = make_shared<TypeParam>(data, target_shape, axes_mapping_const, "EXPLICIT");
+    ASSERT_TRUE(bc->get_output_partial_shape(0).rank().is_dynamic());
+}
+
+TYPED_TEST_P(BroadcastTests, broadcast_explicit_target_shape_rank_static)
```
Comment (arogowie-intel) [inline ngraph/test/type_prop/broadcast.cpp:515]:
```suggestion
TYPED_TEST_P(BroadcastTests, broadcast_explicit_target_shape_and_input_data_rank_static)
```

File: ngraph/test/type_prop/broadcast.cpp [hunk 5]
```diff
@@ -706,5 +1075,114 @@ TEST(type_prop, broadcast_v3_output_rank_deduced_from_new_shape_input)
...
+    ASSERT_EQ(broadcast_v3->get_output_partial_shape(0).rank().get_length(), 5);
+    ASSERT_TRUE(
+        broadcast_v3->get_output_partial_shape(0).same_scheme(PartialShape{8,
+                                                                           Dimension::dynamic(),
+                                                                           Dimension::dynamic(),
+                                                                           Dimension::dynamic(),
+                                                                           Dimension::dynamic()}));
+}
+
+TEST(type_prop, broadcast_v3_bideractional_dynamic_input)
```
Comment (arogowie-intel) [inline ngraph/test/type_prop/broadcast.cpp:1088]:
```suggestion
TEST(type_prop, broadcast_v3_bidirectional_dynamic_input)
```

File: ngraph/test/type_prop/broadcast.cpp [hunk 6]
```diff
@@ -706,5 +1075,114 @@ TEST(type_prop, broadcast_v3_output_rank_deduced_from_new_shape_input)
...
+    broadcast_v3 = make_shared<op::v3::Broadcast>(arg, target_shape, "BIDIRECTIONAL");
+    ASSERT_TRUE(broadcast_v3->get_output_partial_shape(0).rank().is_dynamic());
+
+    // constant target shape
+    const auto target_shape_const = op::Constant::create(element::i64, {3}, {2, 4, 6});
+    broadcast_v3 = make_shared<op::v3::Broadcast>(arg, target_shape_const, "BIDIRECTIONAL");
+    ASSERT_TRUE(broadcast_v3->get_output_partial_shape(0).rank().is_dynamic());
+}
+
+TEST(type_prop, broadcast_v3_bideractional_static_rank_input)
```
Comment (arogowie-intel) [inline ngraph/test/type_prop/broadcast.cpp:1108]:
```suggestion
TEST(type_prop, broadcast_v3_bidirectional_static_rank_input)
```

File: ngraph/test/type_prop/broadcast.cpp [hunk 7]
```diff
@@ -706,5 +1075,114 @@ TEST(type_prop, broadcast_v3_output_rank_deduced_from_new_shape_input)
...
+
+    // constant target shape
+    const auto target_shape_const = op::Constant::create(element::i64, {3}, {2, 4, 6});
+    broadcast_v3 = make_shared<op::v3::Broadcast>(arg, target_shape_const, "BIDIRECTIONAL");
+    ASSERT_TRUE(broadcast_v3->get_output_partial_shape(0).rank().is_static());
+    ASSERT_EQ(broadcast_v3->get_output_partial_shape(0).rank().get_length(), 4);
+    ASSERT_TRUE(broadcast_v3->get_output_partial_shape(0).is_dynamic());
+}
+
+TEST(type_prop, broadcast_v3_bideractional_static_shape_input)
```
Comment (arogowie-intel) [inline ngraph/test/type_prop/broadcast.cpp:1130]:
```suggestion
TEST(type_prop, broadcast_v3_bidirectional_static_shape_input)
```

File: ngraph/test/type_prop/broadcast.cpp [hunk 8]
```diff
@@ -706,5 +1075,114 @@ TEST(type_prop, broadcast_v3_output_rank_deduced_from_new_shape_input)
...
+    // constant target shape
+    const auto target_shape_const = op::Constant::create(element::i64, {4}, {2, 2, 3, 2});
+    broadcast_v3 = make_shared<op::v3::Broadcast>(arg, target_shape_const, "BIDIRECTIONAL");
+    ASSERT_TRUE(broadcast_v3->get_output_partial_shape(0).rank().is_static());
+    ASSERT_EQ(broadcast_v3->get_output_partial_shape(0).rank().get_length(), 4);
+    ASSERT_TRUE(broadcast_v3->get_output_partial_shape(0).is_static());
+    ASSERT_EQ(broadcast_v3->get_output_partial_shape(0), (PartialShape{2, 2, 3, 2}));
+}
+
+TEST(type_prop, broadcast_v3_bideractional_partially_dynamic_input)
```
Comment (arogowie-intel) [inline ngraph/test/type_prop/broadcast.cpp:1153]:
```suggestion
TEST(type_prop, broadcast_v3_bidirectional_partially_dynamic_input)
```

File: ngraph/src/ngraph/op/util/broadcast_base.cpp [hunk 9]
```diff
@@ -46,33 +46,79 @@ op::util::BroadcastBase::BroadcastBase(const Output<Node>& arg,
...
+    for (auto i = start_axis; i < target_shape.size(); i++)
+    {
+        if (arg_shape[i - start_axis].is_dynamic())
+        {
+            continue;
+        }
+        const size_t arg_dim = arg_shape[i - start_axis].get_length();
+        NODE_VALIDATION_CHECK(this,
+                              arg_dim == 1 || arg_dim == target_shape[i],
+                              "Input shape dimension equals ",
```
Comment (arogowie-intel) [inline ngraph/src/ngraph/op/util/broadcast_base.cpp:111]:
```suggestion
                              "Input shape dimension equal ",
```

File: ngraph/test/type_prop/broadcast.cpp [hunk 10]
```diff
@@ -706,5 +1075,114 @@ TEST(type_prop, broadcast_v3_output_rank_deduced_from_new_shape_input)
...
+    auto broadcast_v3 = make_shared<op::v3::Broadcast>(arg, target_shape, "BIDIRECTIONAL");
+    ASSERT_TRUE(broadcast_v3->get_output_partial_shape(0).rank().is_dynamic());
+
+    // static rank target shape
+    target_shape = make_shared<op::Parameter>(element::i64, PartialShape::dynamic(1));
+    broadcast_v3 = make_shared<op::v3::Broadcast>(arg, target_shape, "BIDIRECTIONAL");
+    ASSERT_TRUE(broadcast_v3->get_output_partial_shape(0).rank().is_dynamic());
+
+    // constant target shape
+    const auto target_shape_const = op::Constant::create(element::i64, {4}, {2, 2, 3, 2});
```
Comment (lazarevevgeny) [inline ngraph/test/type_prop/broadcast.cpp:1145]: Lets change that target shape to `{5, 2, 3, 7}`

File: ngraph/src/ngraph/op/broadcast.cpp [hunk 11]
```diff
@@ -88,15 +88,21 @@ std::pair<bool, AxisSet> op::v3::Broadcast::get_broadcast_axes() const
 
 namespace
 {
-    PartialShape
-        get_result_shape_bidirectional(Node* this_ptr, Shape& arg_shape, Shape& target_shape)
+    PartialShape get_result_shape_bidirectional(Node* this_ptr,
+                                                const PartialShape& arg_shape,
+                                                Shape& target_shape)
```
[the remark this answers is not in the document]
Comment (lazarevevgeny) [reply ngraph/src/ngraph/op/broadcast.cpp:93]: Please, disregard the comment. It is opened by mistake.

File: ngraph/test/type_prop/broadcast.cpp (hunk 10 above)
Comment (mbencer) [reply ngraph/test/type_prop/broadcast.cpp]: The test added

File: ngraph/test/eval.cpp [hunk 12]
```diff
@@ -315,7 +315,7 @@ TEST(eval, evaluate_broadcast_v3_numpy_vs_bidi)
     Shape in_shape{1, 4, 1};
 
     auto A = make_shared<op::Parameter>(element::f32, in_shape);
-    auto target_shape = op::Constant::create<int64_t>(element::i64, Shape{3}, {1, 1, 4});
+    auto target_shape = op::Constant::create<int64_t>(element::i64, Shape{3}, {1, 4, 4});
```
[the remark this answers is not in the document]
Comment (mbencer) [reply ngraph/test/eval.cpp:318]: The test added

File: ngraph/test/type_prop/broadcast.cpp [hunk 13]
```diff
@@ -706,5 +1075,121 @@ TEST(type_prop, broadcast_v3_output_rank_deduced_from_new_shape_input)
...
+TEST(type_prop, broadcast_v3_bidirectional_partially_dynamic_input)
+{
+    const auto target_shape =
+        op::Constant::create(element::i64, Shape{4}, vector<int64_t>{1, 1, 50, 50});
+
+    auto data = make_shared<op::Parameter>(element::f32, PartialShape{16, 1, Dimension::dynamic()});
+    auto bc = make_shared<op::v3::Broadcast>(data, target_shape, "BIDIRECTIONAL");
+    ASSERT_TRUE(bc->get_output_partial_shape(0).rank().is_static());
+    ASSERT_EQ(bc->get_output_partial_shape(0).rank().get_length(), 4);
+    ASSERT_EQ(bc->get_output_partial_shape(0), (PartialShape{1, 16, 50, Dimension::dynamic()}));
```
Comment (jane-intel) [inline ngraph/test/type_prop/broadcast.cpp:1169]: We could define the shape fully, aren't we? It should be {1, 16, 50, 50} isn't it?

File: ngraph/test/type_prop/broadcast.cpp [hunk 14]
```diff
@@ -706,5 +1075,121 @@ TEST(type_prop, broadcast_v3_output_rank_deduced_from_new_shape_input)
...
+
+    data = make_shared<op::Parameter>(element::f32,
+                                      PartialShape{Dimension::dynamic(), 1, Dimension::dynamic()});
+    bc = make_shared<op::v3::Broadcast>(data, target_shape, "BIDIRECTIONAL");
+    ASSERT_TRUE(bc->get_output_partial_shape(0).rank().is_static());
+    ASSERT_EQ(bc->get_output_partial_shape(0).rank().get_length(), 4);
+    ASSERT_EQ(bc->get_output_partial_shape(0),
+              (PartialShape{1, Dimension::dynamic(), 50, Dimension::dynamic()}));
```
Comment (jane-intel) [inline ngraph/test/type_prop/broadcast.cpp:1177]: {1, Dimension::dynamic(), 50, 50}?

Comment (jane-intel) [reply ngraph/test/type_prop/broadcast.cpp:1177]: Same for the next tests

File: ngraph/test/type_prop/broadcast.cpp (hunk 13 above)
Comment (mbencer) [reply ngraph/test/type_prop/broadcast.cpp]: Good catch. I think that you are right, but only for scenario when target_dimension != 1 (in such case output dimension can be greater than target dimension).
I relaxed `get_result_shape_bidirectional` method.

File: ngraph/test/type_prop/broadcast.cpp (hunk 14 above)
Comment (mbencer) [reply ngraph/test/type_prop/broadcast.cpp]: as above

**Pull request merged**